ui/client/reducers/graph/utils.ts

```diff
@@ -1,30 +1,51 @@
+import {cloneDeep, map, omit, reject, without, zipWith, curryRight} from "lodash"
+import {Layout, NodePosition, NodesWithPositions} from "../../actions/nk"
+import ProcessUtils from "../../common/ProcessUtils"
 import {ExpressionLang} from "../../components/graph/node-modal/editors/expression/types"
-import {GraphState} from "./types"
-import {NodeType, NodeId, Process, GroupType, Edge, EdgeType, ProcessDefinitionData} from "../../types"
 import NodeUtils from "../../components/graph/NodeUtils"
-import {map, zipWith, omit, cloneDeep, reject} from "lodash"
-import {Layout, NodesWithPositions, NodePosition} from "../../actions/nk"
-import ProcessUtils from "../../common/ProcessUtils"
+import {Edge, EdgeType, GroupType, NodeId, NodeType, Process, ProcessDefinitionData} from "../../types"
+import {GraphState} from "./types"
 
-function isBetween(id1: NodeId, id2: NodeId): (e: Edge) => boolean {
-  return ({from, to}) => from == id1 && to == id2 || from == id2 && to == id1
+function isConnectedTo(...nodeIds: NodeId[]) {
+  return (edge: Edge) => {
+    return nodeIds.find(id => edge.from === id || edge.to === id)
+  }
 }
 
-function canGroup(state: GraphState, newNode: NodeType | GroupType): boolean {
-  const {groupingState, processToDisplay: {edges}} = state
-  const isGroup = NodeUtils.nodeIsGroup(newNode)
-  return !isGroup && groupingState.length === 0 || !!groupingState.find(id => edges.find(isBetween(id, newNode.id)))
+function getSelectedNodes(state: GraphState): NodeType[] {
+  const {processToDisplay, selectionState = []} = state
+  return NodeUtils.nodesFromProcess(processToDisplay).filter(n => selectionState.includes(n.id))
 }
 
-export function displayOrGroup(state: GraphState, node: NodeType, readonly = false): GraphState {
-  if (state.groupingState) {
-    return {
-      ...state,
-      groupingState: canGroup(state, node) ?
-        state.groupingState.concat(node.id) :
-        state.groupingState,
-    }
+export function getSelectedGroups(state: GraphState): GroupType[] {
+  return getSelectedNodes(state).filter(NodeUtils.nodeIsGroup)
+}
+
+export function canGroupSelection(state: GraphState): boolean {
+  const {processToDisplay: {edges}, selectionState = []} = state
+
+  if (selectionState.length < 2) {
+    return false
   }
+
+  const containsGroup = getSelectedGroups(state).length
+  if (containsGroup) {
+    return false
+  }
+
+  const validConnections = getSelectedNodes(state)
+    .map(({id}) => edges
+      .filter(isConnectedTo(id))
+      .filter(isConnectedTo(...without(selectionState, id))))
+    .map(validEgdes => validEgdes.length)
+
+  const lonelyNodes = validConnections.filter(l => l === 0).length
+  const edgeNodes = validConnections.filter(l => l === 1).length
+
+  return !lonelyNodes && edgeNodes === 2
```

Comment:
https://staging.nussknacker.io/visualization/Bounded113?businessView=false - I think `edgeNodes === 2` condition does not allow such group?